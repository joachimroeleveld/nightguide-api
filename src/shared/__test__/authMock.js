const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
