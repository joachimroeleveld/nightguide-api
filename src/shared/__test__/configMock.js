const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
