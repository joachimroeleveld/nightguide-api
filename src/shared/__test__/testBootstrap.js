const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
