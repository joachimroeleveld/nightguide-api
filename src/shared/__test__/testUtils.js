const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
