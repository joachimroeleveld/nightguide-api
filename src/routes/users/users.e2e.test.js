require('../../shared/__test__/testBootstrap');

const request = require('supertest');
const sinon = require('sinon');

const validator = require('../../shared/validator');
const userRepository = require('./userRepository');
const User = require('./userModel');
const { TEST_USER_1, TEST_USER_2 } = require('../../shared/__test__/fixtures');
const { clearDb } = require('../../shared/__test__/testUtils');
const mailService = require('../../shared/services/mail');
const { mockAuth, restoreAuth } = require('../../shared/__test__/authMock');

const sandbox = sinon.createSandbox();

describe('users e2e', () => {
  afterEach(async () => {
    sandbox.restore();
    await clearDb();
  });

  describe('POST /users', () => {
    const validateResponse = validator.validateResponse('post', '/users');

    it('happy path', async () => {
      const res = await request(global.app)
        .post('/users')
        .send(TEST_USER_1);
      const body = res.body;

      expect(res.status).toEqual(201);
      expect(body.email).toEqual(TEST_USER_1.email);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('should set a verification token on the user', async () => {
      await request(global.app)
        .post('/users')
        .send(TEST_USER_1);

      const user = await userRepository.getUserByEmail(
        TEST_USER_1.email,
        '+verificationToken'
      );

      expect(user.verificationToken).toBeTruthy();
    });

    it('should send a verification email', async () => {
      sandbox.stub(mailService, 'sendBasicEmail').resolves();

      await request(global.app)
        .post('/users')
        .send(TEST_USER_1);

      expect(mailService.sendBasicEmail.called).toBe(true);
    });
  });

  describe('GET /users/{userId}', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/users/{userId}'
    );

    it('happy path', async () => {
      const user = await userRepository.createUser(TEST_USER_1, true);

      const res = await request(global.app).get(`/users/${user._id}`);

      expect(res.status).toEqual(200);
      expect(res.body).toMatchInlineSnapshot(
        {
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
        `
Object {
  "createdAt": Any<String>,
  "email": "foo@bar.nl",
  "id": "5c001cac8e84e1067f34695c",
  "updatedAt": Any<String>,
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('only admins can request other users', async () => {
      restoreAuth();

      const user1 = await userRepository.createUser(TEST_USER_1);
      const user2 = await userRepository.createUser(TEST_USER_2);

      const res = await request(global.app)
        .get(`/users/${user2._id}`)
        .set('Authorization', 'Bearer ' + user1.signJwt());

      expect(res.status).toEqual(401);
      expect(validateResponse(res)).toBeUndefined();

      mockAuth();
    });
  });

  describe('POST /users/login', () => {
    const validateResponse = validator.validateResponse('post', '/users/login');

    it('happy path', async () => {
      await userRepository.createUser(TEST_USER_1, true);

      const res = await request(global.app)
        .post('/users/login')
        .send({
          email: TEST_USER_1.email,
          password: TEST_USER_1.password,
        });

      expect(res.status).toEqual(200);
      expect(res.body.user).toMatchInlineSnapshot(
        {
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
        `
Object {
  "createdAt": Any<String>,
  "email": "foo@bar.nl",
  "id": "5c001cac8e84e1067f34695c",
  "updatedAt": Any<String>,
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('returns 404 when user is not found', async () => {
      const res = await request(global.app)
        .post('/users/login')
        .send({
          email: 'foo@bar.nl',
          password: 'pass',
        });

      expect(validateResponse(res)).toBeUndefined();
      expect(res.status).toEqual(404);
    });

    it('returns 401 when passing wrong password', async () => {
      await userRepository.createUser(TEST_USER_1);

      const res = await request(global.app)
        .post('/users/login')
        .send({
          email: TEST_USER_1.email,
          password: 'bar',
        });

      expect(validateResponse(res)).toBeUndefined();
      expect(res.status).toEqual(401);
    });
  });

  describe('GET /users/:userId/verify-account', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/users/{userId}/verify-account'
    );

    it('happy path', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      const res = await request(global.app).get(
        `/users/${user._id}/verify-account?token=${user.verificationToken}`
      );

      const updatedUser = await userRepository.getUserById(
        user._id,
        '+verificationToken'
      );

      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
      expect(res.text).toMatchSnapshot();
      expect(updatedUser.verificationToken).toBeNull();
    });

    it('shows a message when the token is expired', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      // Expire verification token
      user.verificationToken = user.signJwt({}, '0s');
      await user.save();

      const res = await request(global.app).get(
        `/users/${user._id}/verify-account?token=${user.verificationToken}`
      );

      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
      expect(res.text).toEqual(expect.stringMatching('The token has expired'));
    });

    it('returns 401 when token is invalid', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      const res = await request(global.app).get(
        `/users/${user._id}/verify-account?token=invalid-token`
      );

      expect(validateResponse(res)).toBeUndefined();
      expect(res.status).toEqual(401);
    });
  });

  describe('POST /users/:userId/resend-verification-token', () => {
    const validateResponse = validator.validateResponse(
      'post',
      '/users/{userId}/resend-verification-token'
    );

    it('happy path', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      sandbox.stub(user, 'signJwt').returns('testtoken');
      sandbox.stub(userRepository, 'getUserById').resolves(user);

      sandbox.stub(mailService, 'sendBasicEmail').resolves();

      const res = await request(global.app).post(
        `/users/${user._id}/resend-verification-token?token=${
          user.verificationToken
        }`
      );

      expect(res.status).toEqual(200);
      expect(mailService.sendBasicEmail.getCall(0).args).toMatchSnapshot();
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('POST /users/:userId/send-password-reset', () => {
    const validateResponse = validator.validateResponse(
      'post',
      '/users/{userId}/send-password-reset'
    );

    it('happy path', async () => {
      const user = new User(TEST_USER_1);

      sandbox.stub(user, 'signJwt').returns('testtoken');
      sandbox.stub(userRepository, 'getUserById').resolves(user);

      sandbox.stub(mailService, 'sendBasicEmail').resolves();

      const res = await request(global.app).post(
        `/users/${user._id}/send-password-reset`
      );

      expect(res.status).toEqual(200);
      expect(mailService.sendBasicEmail.getCall(0).args).toMatchSnapshot();
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /users/:userId/reset-password', () => {
    const validateResponse = validator.validateResponse(
      'get',
      '/users/{userId}/reset-password'
    );

    it('happy path', async () => {
      const user = await userRepository.createUser({
        ...TEST_USER_1,
        password: 'test-password',
        passwordResetToken: 'testtoken',
      });

      const res = await request(global.app).get(
        `/users/${user._id}/reset-password?token=testtoken`
      );

      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
      expect(res.text).toMatchSnapshot();
    });
  });

  describe('POST /users/:userId/reset-password', () => {
    const validateResponse = validator.validateResponse(
      'post',
      '/users/{userId}/reset-password'
    );

    it('happy path', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      user.passwordResetToken = user.signJwt({}, '1h');
      await user.save();

      const NEW_PASSWORD = 'new-password';

      const res = await request(global.app)
        .post(
          `/users/${user._id}/reset-password?token=${user.passwordResetToken}`
        )
        .send({
          password: NEW_PASSWORD,
        });

      expect(res.status).toBe(200);
      expect(validateResponse(res)).toBeUndefined();

      const updatedUser = await userRepository.getUserById(user._id);
      expect(await updatedUser.verifyPassword(NEW_PASSWORD)).toBe(true);
    });

    it('returns 401 when the token is expired', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      // Expire verification token
      user.passwordResetToken = user.signJwt({}, '0s');
      await user.save();

      const res = await request(global.app)
        .post(
          `/users/${user._id}/reset-password?token=${user.passwordResetToken}`
        )
        .send({
          password: 'foo',
        });

      expect(validateResponse(res)).toBeUndefined();
      expect(res.status).toEqual(401);
    });

    it('returns 401 when token is invalid', async () => {
      const user = await userRepository.createUser(TEST_USER_1);

      const res = await request(global.app)
        .post(`/users/${user._id}/reset-password?token=invalid-token`)
        .send({
          password: 'foo',
        });

      expect(validateResponse(res)).toBeUndefined();
      expect(res.status).toEqual(401);
    });
  });
});
