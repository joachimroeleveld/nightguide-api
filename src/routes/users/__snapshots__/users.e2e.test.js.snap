// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users e2e GET /users/:userId/reset-password happy path 1`] = `
"<html>
  <head>
    <title>NightGuide password reset</title>
  </head>
  <body>
    <style>
      html, body {
        background: #1f1f1f;
        font-family: 'Helvetica Neue', Helvetica, Arial;
      }
      h1 {
        color: #fff;
        margin-top: 2em;
      }
      main {
        max-width: 860px;
        padding: 1em;
        margin: 0 auto;
      }
      button {
        background: none;
        border: none;
        background: #2BAADF;
        color: #fff;
        padding: 1em;
        margin: 1em 0;
        border-radius: 2px;
        font-size: 1rem;
      }
      form {
        padding: 0.1em 0;
      }
      .form-item {
        margin: 1em 0;
        color: #fff;
      }
      .form-item label {
        display: block;
        margin-bottom: 0.3em;
      }
      input[type=\\"text\\"], input[type=\\"password\\"] {
        background: rgba(255,255,255,0.2);
        border: none;
        padding: 0.5em 1em;
        border-radius: 2px;
        min-width: 300px;
        color: #fff;
        font-size: 1rem;
      }
    </style>
    <main>
      <p>
        <img alt=\\"NightGuide logo\\" src=\\"https://storage.googleapis.com/static.nightguide.app/logo-184x97.png\\" width=\\"183\\" style=\\"max-width:183px; padding-bottom: 0; display: inline !important; vertical-align: bottom;\\" class=\\"mcnImage\\">
      </p>
      <p>
        <h1>Reset your password</h1>

        <form>
          <div class=\\"form-item\\">
            <label for=\\"password\\">New password</label>
            <input name=\\"password\\" id=\\"password\\" type=\\"password\\">
          </div>
          <div class=\\"form-item\\">
            <label for=\\"password\\">Repeat password</label>
            <input id=\\"password-repeat\\" type=\\"password\\">
          </div>
        </form>

        <button id=\\"reset-submit\\">Reset password</button>

        <script src=\\"https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js\\"></script>
        <script>
          (function($) {
            let loading = false;

            $('#reset-submit').click(function() {
              if (loading) return;

              loading = true;

              if ($('#password').value !== $('#password-repeat').value) {
                alert('The passwords do not match')
                return;
              }

              $.ajax({
                type: 'POST',
                url: 'http://localhost:8080/users/5c001cac8e84e1067f34695c/reset-password?token=testtoken',
                data: convertFormToJSON('form'),
                contentType: 'application/json',
              })
                .done(function() {
                  $('h1').text('Your password has been reset');
                  $('form, button').remove();
                })
                .fail(function(e) {
                  $('h1').text('Something went wrong, try again');
                  loading = false;
                });
            });

            function convertFormToJSON(form){
              var array = jQuery(form).serializeArray();
              var obj = {};

              jQuery.each(array, function() {
                obj[this.name] = this.value || '';
              });

              return JSON.stringify(obj);
            }
          })(jQuery)
        </script>
      </p>
    </main>
  </body>
</html>
"
`;

exports[`users e2e GET /users/:userId/verify-account happy path 1`] = `
"<html>
  <head>
    <title>NightGuide account verified</title>
  </head>
  <body>
    <style>
      html, body {
        background: #1f1f1f;
        font-family: 'Helvetica Neue', Helvetica, Arial;
      }
      h1 {
        color: #fff;
        margin-top: 2em;
      }
      main {
        max-width: 860px;
        padding: 1em;
        margin: 0 auto;
      }
      button {
        background: none;
        border: none;
        background: #2BAADF;
        color: #fff;
        padding: 1em;
        margin: 1em 0;
        border-radius: 2px;
        font-size: 1rem;
      }
    </style>
    <main>
      <p>
        <img alt=\\"NightGuide logo\\" src=\\"https://storage.googleapis.com/static.nightguide.app/logo-184x97.png\\" width=\\"183\\" style=\\"max-width:183px; padding-bottom: 0; display: inline !important; vertical-align: bottom;\\" class=\\"mcnImage\\">
      </p>
      <p>
          <h1>Your account has been verified</h1>
      </p>
    </main>
  </body>
</html>
"
`;

exports[`users e2e POST /users happy path 1`] = `
Object {
  "birthday": "1995-10-04",
  "createdAt": Any<String>,
  "email": "alice@rogers.nl",
  "firstName": "Alice",
  "gender": "female",
  "id": Any<String>,
  "lastName": "Rogers",
  "updatedAt": Any<String>,
}
`;

exports[`users e2e POST /users updates the user if another authentication method was used before 1`] = `
Object {
  "birthday": "1995-10-04",
  "createdAt": Any<String>,
  "email": "facebook@roy.com",
  "firstName": "Alice",
  "gender": "female",
  "id": Any<String>,
  "lastName": "Rogers",
  "updatedAt": Any<String>,
}
`;

exports[`users e2e POST /users/:userId/resend-verification-token happy path 1`] = `
Array [
  "alice@rogers.nl",
  "Verify your account",
  "
You can verify your account by following <a href=\\"http://localhost:8080/users/5c001cac8e84e1067f34695c/verify-account?token=testtoken\\" target=\\"_blank\\">this link</a>. After you verified your account, you can login and start using your account.<br/>
<br/>
<a href=\\"http://localhost:8080/users/5c001cac8e84e1067f34695c/verify-account?token=testtoken\\" target=\\"_blank\\">http://localhost:8080/users/5c001cac8e84e1067f34695c/verify-account?token=testtoken</a>
    ",
]
`;

exports[`users e2e POST /users/login happy path 1`] = `
Object {
  "birthday": "1995-10-04",
  "createdAt": Any<String>,
  "email": "alice@rogers.nl",
  "firstName": "Alice",
  "gender": "female",
  "id": Any<String>,
  "lastName": "Rogers",
  "updatedAt": Any<String>,
}
`;

exports[`users e2e POST /users/login-fb happy path 1`] = `
Object {
  "birthday": "1995-10-04",
  "createdAt": Any<String>,
  "email": "alice@rogers.nl",
  "firstName": "Alice",
  "gender": "female",
  "id": Any<String>,
  "lastName": "Rogers",
  "updatedAt": Any<String>,
}
`;

exports[`users e2e POST /users/send-password-reset happy path 1`] = `
Array [
  "alice@rogers.nl",
  "Reset your password",
  "
You can reset your password by following <a href=\\"http://localhost:8080/users/5c001cac8e84e1067f34695c/reset-password?token=testtoken\\" target=\\"_blank\\">this link</a>.<br/>
<br/>
<a href=\\"http://localhost:8080/users/5c001cac8e84e1067f34695c/reset-password?token=testtoken\\" target=\\"_blank\\">http://localhost:8080/users/5c001cac8e84e1067f34695c/reset-password?token=testtoken</a>
    ",
]
`;
